scripts/check_cmake_files.py

```diff
@@ -0,0 +1,282 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Checks that source files are mentioned in CMakeLists.txt."""
+
+import argparse
+import collections
+import os
+import re
+import subprocess
+import sys
+
+
+DEFAULT_DIRS = [
+    'Firestore/core',
+    'Firestore/Source'
+]
+
+
+_verbose = False
+
+
+def main(args):
+  global _verbose
+
+  parser = argparse.ArgumentParser(description='Lint source files.')
+  parser.add_argument('--verbose', '-v', action='store_true',
+                      help='Run verbosely')
+  parser.add_argument('filenames', nargs='*', metavar='filename',
+                      help='Filenames to scan')
+  args = parser.parse_args(args)
+  if args.verbose:
+    _verbose = True
+
+  scan_filenames = args.filenames
+  if len(scan_filenames) == 0:
+    scan_filenames = default_args()
+
+  filenames = find_source_files(scan_filenames)
+  groups = group_by_cmakelists(filenames)
+  errors = find_all_errors(groups)
+  trace('checked %d files' % len(filenames))
+  sys.exit(1 if errors else 0)
+
+
+def default_args():
+  """Returns a default list of directories to scan.
+  """
+  command = ['git', 'rev-parse', '--show-toplevel']
+  toplevel = subprocess.check_output(command).rstrip()
+
+  return [os.path.join(toplevel, dirname) for dirname in DEFAULT_DIRS]
+
+
+def find_source_files(roots):
+  """Finds source files on the filesystem.
+
+  Args:
+    roots: A list of files or directories
+
+  Returns:
+    A list of filenames found in the roots, excluding those that are
+    uninteresting.
+
+  """
+  result = []
+  exclude_dirs = {'third_party', 'Pods', 'Protos'}
+  include_files = {'CMakeLists.txt'}
+
+  for root in roots:
+    for parent, dirs, files in os.walk(root, topdown=True):
+      # Prune directories known to be uninteresting
+      dirs[:] = [d for d in dirs if d not in exclude_dirs]
+
+      for filename in files:
+        if filename in include_files or is_source_file(filename):
+          result.append(os.path.join(parent, filename))
+
+  return result
+
+
+_filename_pattern = re.compile(r'\b([A-Za-z0-9_/+]+\.)+(?:c|cc|h|m|mm)\b')
+_check_pattern = re.compile(r'^\s*check_[A-Za-z0-9_]+\(.*\)$')
+_ignore_pattern = re.compile(r'GENERATED')
+
+
+def read_listed_source_files(filename):
+  """Reads the contents of the given filename and finds all the filenames it
+  finds in the file.
+
+  Args:
+    filename: A filename to read, typically some path to a CMakeLists.txt file.
+
+  Returns:
+    A pair of lists. The first list is the list of filenames mentioned in the
+    file. The second is a list of files that have been marked GENERATED in the
+    file.
+  """
+  found = []
+  ignored = []
+  parent = os.path.dirname(filename)
+
+  with open(filename, 'r') as fd:
+    for line in fd.readlines():
+      # Simple hack to exclude files mentioned in CMake checks
```

Comment:
This was sort of a feature--it was this script that caused me to add those commented out files in the first place. However, after thinking about it, I think requiring those files to be marked NOLINT is probably a better idea.

I've added comment handling.

Reply:
Honestly, I'd go with the simpler approach:
* ignore all comments;
* always require non-commented-out files that have to be ignored to be annotated with `NOLINT` (even if it requires annotating the same file more than once). I actually cannot think of a good reason for such a `NOLINT` file to appear more than once in the CMake files.

I don't feel strongly about this, so feel free to keep as is.

Reply:
The trouble is that there are two kinds of files to mark NOLINT

  * Files that are generated, and are therefore visible to CMake
  * Files that aren't handled in CMake yet

It's this second group that are tricky to handle because, by virtue of being unhandled, they can't actually be visible to CMake. An easy way to handle this is with a comment explaining that the file is unhandled, and NOLINT that mention. If I unconditionally strip comments this path is longer viable.

I'm not against doing something simpler but I don't have a great alternative for handling files of this second type. What's your proposal?

Reply:
Never mind, now that I think of it, what I had in mind is infeasible.